File: src/neetcode/Java/_10RegularExpressionMatching.java
package neetcode.Java;

public class _10RegularExpressionMatching {
    String ss = "", pp = "";
    int s_len, p_len;

    public boolean isMatch(String s, String p) {
        ss = s;
        pp = p;
        s_len = s.length();
        p_len = p.length();
        return Find(0, 0);
    }

    public boolean Find(int i, int j) {
        if (ss.length() == i && pp.length() == j) return true;
        if (ss.length() != i && pp.length() == j) return false;
        if (j + 1 < p_len && pp.charAt(j + 1) == '*') {
            if (i < s_len && (pp.charAt(j) == '.' || ss.charAt(i) == pp.charAt(j))) {
                return (Find(i, j + 2) || Find(i + 1, j));
            } else {
                return Find(i, j + 2);
            }
        }
        if (i < s_len && (pp.charAt(j) == '.' || ss.charAt(i) == pp.charAt(j)))
            return Find(i + 1, j + 1);
        return false;
    }


    public static void main(String[] args) {
        _10RegularExpressionMatching _10RegularExpressionMatching = new _10RegularExpressionMatching();
        boolean res = _10RegularExpressionMatching.isMatch("aa", ".*");
        System.out.println(res);
    }
}

File: src/neetcode/kotlin/_10RegularExpressionMatching.kt
package neetcode.kotlin

class _10RegularExpressionMatching {
    var ss: String = ""
    var pp: String = ""
    var s_len: Int = 0
    var p_len: Int = 0

    fun isMatch(s: String, p: String): Boolean {
        ss = s
        pp = p
        s_len = s.length
        p_len = p.length
        return Find(0, 0)
    }

    fun Find(i: Int, j: Int): Boolean {
        if (ss.length == i && pp.length == j) return true
        if (ss.length != i && pp.length == j) return false
        if (j + 1 < p_len && pp[j + 1] == '*') {
            return if (i < s_len && (pp[j] == '.' || ss[i] == pp[j])) {
                Find(i, j + 2) || Find(i + 1, j)
            } else {
                Find(i, j + 2)
            }
        }
        if (i < s_len && (pp[j] == '.' || ss[i] == pp[j])) return Find(i + 1, j + 1)
        return false
    }


    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val _10RegularExpressionMatching: _10RegularExpressionMatching =
                neetcode.kotlin._10RegularExpressionMatching()
            val res = _10RegularExpressionMatching.isMatch("aa", "a")
            println(res)
        }
    }
}
